File: src/aica_backend/utils/validation.py
import re
from typing import Tuple, Optional, Dict, Any
from datetime import datetime, date


class ValidationUtils:
    EMAIL_PATTERN = re.compile(
        r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    )
    PHONE_PATTERN = re.compile(
        r'^\+?[\d\s\-\(\)]+$'
    )
    URL_PATTERN = re.compile(
        r'^https?:\/\/(?:[-\w.])+(?:\:[0-9]+)?(?:\/(?:[\w\/_.])*(?:\?(?:[\w&=%.])*)?(?:\#(?:[\w.])*)?)?$'
    )
    
    @classmethod
    def validate_email(cls, email: str) -> Tuple[bool, Optional[str]]:
        if not email:
            return False, "Email is required"
        
        if not cls.EMAIL_PATTERN.match(email):
            return False, "Invalid email format"
        
        if len(email) > 254: 
            return False, "Email address too long"
        
        return True, None
    
    @classmethod
    def validate_phone(cls, phone: str) -> Tuple[bool, Optional[str]]:
        if not phone:
            return False, "Phone number is required"
        
        clean_phone = re.sub(r'\s', '', phone)
        
        if len(clean_phone) < 10:
            return False, "Phone number too short"
        
        if len(clean_phone) > 15:  
            return False, "Phone number too long"
        
        if not cls.PHONE_PATTERN.match(phone):
            return False, "Invalid phone number format"
        
        return True, None
    
    @classmethod
    def validate_url(cls, url: str) -> Tuple[bool, Optional[str]]:
        if not url:
            return True, None 
        
        if not cls.URL_PATTERN.match(url):
            return False, "Invalid URL format"
        
        if len(url) > 2048:  # Common browser limit
            return False, "URL too long"
        
        return True, None
    
    @classmethod
    def validate_date_range(cls, start_date: date, end_date: Optional[date]) -> Tuple[bool, Optional[str]]:
        if not start_date:
            return False, "Start date is required"
        
        if start_date > date.today():
            return False, "Start date cannot be in the future"
        
        if end_date:
            if end_date < start_date:
                return False, "End date must be after start date"
            
            if end_date > date.today():
                return False, "End date cannot be in the future"
        
        return True, None
    
    @classmethod
    def validate_text_length(cls, text: str, field_name: str, min_length: int = 0, max_length: int = 1000) -> Tuple[bool, Optional[str]]:
        if not text and min_length > 0:
            return False, f"{field_name} is required"
        
        if text and len(text.strip()) < min_length:
            return False, f"{field_name} must be at least {min_length} characters"
        
        if text and len(text) > max_length:
            return False, f"{field_name} must not exceed {max_length} characters"
        
        return True, None
    
    @classmethod
    def validate_profile_data(cls, profile_data: Dict[str, Any]) -> Tuple[bool, Dict[str, str]]:
        errors = {}
    
        required_fields = {
            'first_name': (1, 100),
            'last_name': (1, 100),
            'professional_title': (1, 200),
            'summary': (10, 2000),
        }
        
        for field, (min_len, max_len) in required_fields.items():
            value = profile_data.get(field, '')
            is_valid, error = cls.validate_text_length(value, field.replace('_', ' ').title(), min_len, max_len)
            if not is_valid:
                errors[field] = error
        
        if 'email' in profile_data:
            is_valid, error = cls.validate_email(profile_data['email'])
            if not is_valid:
                errors['email'] = error
        
        if 'contact_number' in profile_data:
            is_valid, error = cls.validate_phone(profile_data['contact_number'])
            if not is_valid:
                errors['contact_number'] = error
        
        if 'linkedin_url' in profile_data:
            is_valid, error = cls.validate_url(profile_data['linkedin_url'])
            if not is_valid:
                errors['linkedin_url'] = error
        
        address = profile_data.get('address', '')
        is_valid, error = cls.validate_text_length(address, 'Address', 5, 500)
        if not is_valid:
            errors['address'] = error
        
        return len(errors) == 0, errors
    
    @classmethod
    def validate_experience_data(cls, experience_data: Dict[str, Any]) -> Tuple[bool, Dict[str, str]]:
        errors = {}
        
        job_title = experience_data.get('job_title', '')
        is_valid, error = cls.validate_text_length(job_title, 'Job Title', 1, 200)
        if not is_valid:
            errors['job_title'] = error
        
        company_name = experience_data.get('company_name', '')
        is_valid, error = cls.validate_text_length(company_name, 'Company Name', 1, 200)
        if not is_valid:
            errors['company_name'] = error
        
        start_date = experience_data.get('start_date')
        end_date = experience_data.get('end_date')
        is_current = experience_data.get('is_current', False)
        
        if start_date:
            if isinstance(start_date, str):
                try:
                    start_date = datetime.strptime(start_date, '%Y-%m-%d').date()
                except ValueError:
                    errors['start_date'] = "Invalid date format, use YYYY-MM-DD"
                    start_date = None
            
            if end_date and isinstance(end_date, str):
                try:
                    end_date = datetime.strptime(end_date, '%Y-%m-%d').date()
                except ValueError:
                    errors['end_date'] = "Invalid date format, use YYYY-MM-DD"
                    end_date = None
                    
            if start_date and not is_current:
                is_valid, error = cls.validate_date_range(start_date, end_date)
                if not is_valid:
                    errors['date_range'] = error
        else:
            errors['start_date'] = "Start date is required"
        
        return len(errors) == 0, errors

File: src/aica_frontend/lib/schemas/validation.ts
import { z } from 'zod';

export const commonValidations = {
  email: z
    .string()
    .min(1, 'Email is required')
    .email('Please enter a valid email address')
    .max(254, 'Email is too long'),

  password: z
    .string()
    .min(8, 'Password must be at least 8 characters long')
    .max(128, 'Password is too long')
    .regex(/[A-Z]/, 'Password must contain at least one uppercase letter')
    .regex(/[a-z]/, 'Password must contain at least one lowercase letter')
    .regex(/[0-9]/, 'Password must contain at least one number')
    .regex(
      /[!@#$%^&*(),.?":{}|<>]/,
      'Password must contain at least one special character',
    ),

  confirmPassword: z.string().min(1, 'Please confirm your password'),

  firstName: z
    .string()
    .min(2, 'First name must be at least 2 characters')
    .max(50, 'First name must be less than 50 characters')
    .regex(
      /^[A-Za-z\s\-']+$/,
      'First name should only contain letters, spaces, hyphens, and apostrophes',
    ),

  lastName: z
    .string()
    .min(2, 'Last name must be at least 2 characters')
    .max(50, 'Last name must be less than 50 characters')
    .regex(
      /^[A-Za-z\s\-']+$/,
      'Last name should only contain letters, spaces, hyphens, and apostrophes',
    ),

  phoneNumber: z
    .string()
    .min(7, 'Invalid contact number')
    .max(20, 'Contact number is too long')
    .regex(/^[\+\d\s\-\(\)]+$/, 'Please enter a valid phone number'),

  url: z.string().url('Please enter a valid URL').optional().or(z.literal('')),

  requiredString: (fieldName: string, minLength = 1, maxLength = 100) =>
    z
      .string()
      .min(minLength, `${fieldName} is required`)
      .max(maxLength, `${fieldName} must be less than ${maxLength} characters`),

  optionalString: (maxLength = 100) =>
    z
      .string()
      .max(maxLength, `Must be less than ${maxLength} characters`)
      .optional(),

  dateString: z
    .string()
    .min(1, 'Date is required')
    .regex(
      /^\d{4}-\d{2}-\d{2}$/,
      'Please enter a valid date in YYYY-MM-DD format',
    ),

  optionalDateString: z
    .string()
    .regex(
      /^\d{4}-\d{2}-\d{2}$/,
      'Please enter a valid date in YYYY-MM-DD format',
    )
    .optional()
    .or(z.literal('')),
};

// Authentication schemas
export const loginSchema = z.object({
  email: commonValidations.email,
  password: z.string().min(1, 'Password is required'),
});

export const registerSchema = z
  .object({
    email: commonValidations.email,
    password: commonValidations.password,
    confirmPassword: commonValidations.confirmPassword,
  })
  .refine(data => data.password === data.confirmPassword, {
    message: "Passwords don't match",
    path: ['confirmPassword'],
  });

// Profile schemas
export const profileSchema = z.object({
  first_name: commonValidations.firstName,
  last_name: commonValidations.lastName,
  professional_title: commonValidations.requiredString('Professional title'),
  contact_number: commonValidations.phoneNumber,
  address: commonValidations.optionalString(255),
  linkedin_url: commonValidations.url,
  summary: commonValidations.optionalString(1000),
});

// Skills schema
export const skillSchema = z.object({
  name: commonValidations.requiredString('Skill name', 1, 50),
});

export const skillsSchema = z.object({
  skills: z.array(skillSchema).min(1, 'At least one skill is required'),
});

// Education schemas
export const educationSchema = z.object({
  institution_name: commonValidations.requiredString(
    'Institution name',
    2,
    100,
  ),
  degree: commonValidations.requiredString('Degree', 2, 100),
  field_of_study: commonValidations.optionalString(100),
  address: commonValidations.requiredString('Address', 2, 255),
  start_date: commonValidations.dateString,
  end_date: commonValidations.dateString,
});

export const educationsSchema = z.object({
  educations: z
    .array(educationSchema)
    .min(1, 'At least one education entry is required'),
});

// Experience schemas
export const experienceSchema = z.object({
  job_title: commonValidations.requiredString('Job title', 2, 100),
  company_name: commonValidations.requiredString('Company name', 2, 100),
  start_date: commonValidations.dateString,
  end_date: commonValidations.optionalDateString,
  description: z
    .array(z.string().min(1, 'Description cannot be empty'))
    .min(1, 'At least one description is required')
    .optional(),
  is_current: z.boolean().default(false),
});

export const experiencesSchema = z.object({
  experiences: z.array(experienceSchema).optional().default([]),
});

// Certificate schemas
export const certificateSchema = z.object({
  name: commonValidations.optionalString(100),
  issuing_organization: commonValidations.optionalString(100),
  issue_date: commonValidations.optionalDateString,
  credential_url: commonValidations.url,
  credential_id: commonValidations.optionalString(50),
});

export const certificatesSchema = z.object({
  certificates: z.array(certificateSchema),
});

// Password change schema
export const passwordChangeSchema = z
  .object({
    currentPassword: z.string().min(1, 'Current password is required'),
    newPassword: commonValidations.password,
    confirmNewPassword: commonValidations.confirmPassword,
  })
  .refine(data => data.newPassword === data.confirmNewPassword, {
    message: "Passwords don't match",
    path: ['confirmNewPassword'],
  });

// Export types for TypeScript
export type LoginFormData = z.infer<typeof loginSchema>;
export type RegisterFormData = z.infer<typeof registerSchema>;
export type ProfileFormData = z.infer<typeof profileSchema>;
export type SkillsFormData = z.infer<typeof skillsSchema>;
export type EducationsFormData = z.infer<typeof educationsSchema>;
export type ExperiencesFormData = z.infer<typeof experiencesSchema>;
export type CertificatesFormData = z.infer<typeof certificatesSchema>;
export type PasswordChangeFormData = z.infer<typeof passwordChangeSchema>;
